frontend/src/GameRouter.tsx

```diff
@@ -1,29 +1,76 @@
-import React from "react";
-import {
-  MemoryRouter,
-  Switch,
-  Route,
-  Redirect,
-  useParams,
-} from "react-router-dom";
+import React, { useContext, useState } from "react";
+import { Router, Switch, Route, Redirect, useParams } from "react-router-dom";
+import { createMemoryHistory } from "history";
 import { BasicPage, NicknamePage, LobbyPage } from "./pages";
+import socket from "./socket";
+import {
+  Player,
+  PlayersContext,
+} from "./ContextProviders/PlayersContextProvider";
+
+export type Settings = {
+  roundLimit?: number;
+  maxPlayers?: number;
+};
 
 type PathParams = {
   gameCode: string;
 };
 
+const memoryHistory = createMemoryHistory();
+
 const GameRouter = () => {
   const { gameCode } = useParams<PathParams>();
+  const { setHost, initialisePlayers, addPlayer, removePlayer } = useContext(
+    PlayersContext
+  );
+  const [settings, setSettings] = useState<Settings>();
+
+  socket.on("navigate", (route: string) => memoryHistory.push(route));
+  socket.on("host", (host: string) => setHost(host));
+  socket.on("players:initial", (players: Player[]) =>
+    initialisePlayers(players)
+  );
+  socket.on("players:add", (id: string, nickname: string) =>
+    addPlayer(id, nickname)
+  );
+  socket.on("players:remove", (playerId: string) => removePlayer(playerId));
+  socket.on("settings:initial", (initialSettings: Settings) =>
+    setSettings(initialSettings)
+  );
+  socket.on(
+    "settings:update",
+    ({ setting, value }: { setting: string; value: number }) => {
+      let key;
+      switch (setting) {
+        case "MAX_PLAYERS":
+          key = "maxPlayers";
+          break;
+        case "ROUND_LIMIT":
+          key = "roundLimit";
+          break;
+        default:
+          key = null;
+      }
+      if (key !== null) {
+        const newSettings = {
+          ...settings,
+          [key]: value,
+        };
+        setSettings(newSettings);
+      }
+    }
+  );
```

Comment:
We need to bind these listeners inside a useEffect(), alongside unbinding when the component unmounts. An example of this can be seen here: [https://dev.to/bravemaster619/how-to-use-socket-io-client-correctly-in-react-app-o65](https://dev.to/bravemaster619/how-to-use-socket-io-client-correctly-in-react-app-o65)

Reply:
Similarly it seems that `useCallback()` should be used for performance improvements too